internal/nitrite.go

```diff
@@ -0,0 +1,415 @@
+// Result will be non-null if and only if either of these are true: certificate
+// verification has passed, certificate verification has failed (expired, not
+// trusted, etc.), signature is OK or signature is not OK. If either signature
+// is not OK or certificate can't be verified, both Result and error will be
+// set! You can use the SignatureOK field from the result to distinguish
+// errors.
+func Verify(
+	attestation []byte,
+	certProvider CertProvider,
+	verificationTime VerificationTimeFunc,
+	allowDebug bool,
+) (
+	*Result,
+	error,
+) {
+	cose := cosePayload{}
+
+	err := cbor.Unmarshal(attestation, &cose)
+	if nil != err {
+		return nil, ErrBadCOSESign1Structure
+	}
+
+	if nil == cose.Protected || 0 == len(cose.Protected) {
+		return nil, ErrCOSESign1EmptyProtectedSection
+	}
+
+	if nil == cose.Payload || 0 == len(cose.Payload) {
+		return nil, ErrCOSESign1EmptyPayloadSection
+	}
+
+	if nil == cose.Signature || 0 == len(cose.Signature) {
+		return nil, ErrCOSESign1EmptySignatureSection
+	}
+
+	header := CoseHeader{}
+	err = cbor.Unmarshal(cose.Protected, &header)
+	if nil != err {
+		return nil, ErrBadCOSESign1Structure
+	}
+
+	intAlg, ok := header.AlgorithmInt()
+
+	// https://datatracker.ietf.org/doc/html/rfc8152#section-8.1
+	if ok {
+		switch intAlg {
+		case -35:
+			// do nothing -- OK
+
+		default:
+			return nil, ErrCOSESign1BadAlgorithm
+		}
+	} else {
+		strAlg, ok := header.AlgorithmString()
+
+		if ok {
+			switch strAlg {
+			case "ES384":
+				// do nothing -- OK
+
+			default:
+				return nil, ErrCOSESign1BadAlgorithm
+			}
+		} else {
+			return nil, ErrCOSESign1BadAlgorithm
+		}
+	}
+
+	doc := Document{}
+
+	err = cbor.Unmarshal(cose.Payload, &doc)
+	if nil != err {
+		return nil, ErrBadAttestationDocument
+	}
+
+	if "" == doc.ModuleID ||
+		"" == doc.Digest ||
+		0 == doc.Timestamp ||
+		nil == doc.PCRs ||
+		nil == doc.Certificate {
+		return nil, ErrMandatoryFieldsMissing
+	}
+
+	cert, err := x509.ParseCertificate(doc.Certificate)
+	if nil != err {
+		return nil, err
+	}
+
+	// TODO: remove the support for self-signed attestations as part of
+	//  https://blocky.atlassian.net/browse/BKY-5620 (remove !cert.IsCA path)
+	if !cert.IsCA && nil == doc.CABundle {
+		return nil, ErrMandatoryFieldsMissing
+	}
+
+	if "SHA384" != doc.Digest {
+		return nil, ErrBadDigest
+	}
+
+	if doc.Timestamp < 1 {
+		return nil, ErrBadTimestamp
+	}
+
+	if len(doc.PCRs) < 1 || len(doc.PCRs) > 32 {
+		return nil, ErrBadPCRs
+	}
+
+	for key, value := range doc.PCRs {
+		if key > 31 {
+			return nil, ErrBadPCRIndex
+		}
+
+		if nil == value ||
+			!(32 == len(value) ||
+				48 == len(value) ||
+				64 == len(value) ||
+				96 == len(value)) {
+			return nil, ErrBadPCRValue
+		}
+	}
+
+	docDebug, err := doc.Debug()
+	if err != nil {
+		return nil, fmt.Errorf("getting document debug: %w", err)
+	}
+
+	if !allowDebug && docDebug {
+		return nil, fmt.Errorf("attestation was generated in debug mode")
+	}
```

Comment:
In addition to adding `Document.Debug()` we're also checking it in `internal.Verify()` (moved from `nitrite.Verify`)